lib/Service/CapabilitiesService.php

```diff
@@ -113,6 +113,7 @@ public function refetch() {
 		}
 
 		$this->capabilities = $capabilities;
-		$this->cache->set('capabilities', $capabilities, 3600);
+		if (count($capabilities) !== 0)
+			$this->cache->set('capabilities', $capabilities, 3600);
```

Comment:
I'm a bit concerned here about that this will cause every request to fetch the capabilities again so in case there is some issue where the request times out or takes very long this would add to all the requests where nextcloud tries to obtain the capabilities. How about just storing it with a much lower ttl in the cache in case of the exception above?

Reply:
I think it will be good compromise, what ttl do you propose? 60 s ?

Reply:
Yep, sounds sane to me :+1:

Reply:
@juliushaertl What is the purpose of refetching the capabilities at this point? I'd assume, if there are connectivity problems, a timeout would be the expected behavior, and the already cached capabilities are suitable for that. Is that not the case?

The exception I can think of is if the server is being reconfigured, which is probably done very rarely,

Reply:
@thebearon when capabilities are needed first there is a cache used (see `getCapabilities`), but when we don't have anything in the cache then `refetch` function is executed.
Exception above my code is fired for example when Collabora Online server is not started at all. When server is re/starting then from my experience we don't receive any exception here, but normal response (after some longer time when capabilities enpoint is ready).

Reply:
> @thebearon when capabilities are needed first there is a cache used (see `getCapabilities`), but when we don't have anything in the cache then `refetch` function is executed.

@eszkadev Right, but in my repro steps I opened online just a minute before, thus the capabilities must've gotten cached. Then I turn online off to simulate a connectivity problem, I try opening a file again, and somewhere around this point it refetches the empty capabilities, because when I try once again afterwards (with online available again), the cached "good" capabilities are gone. This refetch seems incorrect to me.

The fix of lowering the lifetime of cached empty capabilities is good, just an additional piece of fix is needed in my opinion, unless there's a sound reason why that refetch happens (obviously I don't know much about what's happening there ;)).

Reply:
> @eszkadev Right, but in my repro steps I opened online just a minute before, thus the capabilities must've gotten cached. Then I turn online off to simulate a connectivity problem, I try opening a file again, and somewhere around this point it refetches the empty capabilities, because when I try once again afterwards (with online available again), the cached "good" capabilities are gone. This refetch seems incorrect to me.

Do you have APCu/Redis caching setup on your instance? Otherwise it would fetch them always anyways

Reply:
@thebearon For me it works (it remembers capabilities and opens editor, but receives 404 because Collabora is off), no refetching happens and no error. I cannot reproduce your steps

Reply:
> Do you have APCu/Redis caching setup on your instance? Otherwise it would fetch them always anyways

@juliushaertl I haven't set up anything special on my instance, so I presume I don't have those. Yet I'm still having this issue.

Reply:
Can you quickly check your nextclouds config.php for entries mentioning `memcache.local` and `memcache.distributed`?

Reply:
@juliushaertl I have neither in my config.php.

Reply:
OK, strange as then the capabilities should be refetched on every page load.

Just to make sure we don't miss something obvious here, do you reproduction steps include to reload the page after stopping/starting the online server? If not, that might be something for the additional fix mentioned via email for checking each time before opening a document.

Reply:
The steps do include reloading the page after stopping the server, because you get stuck on a loading screen (unless you wait that out), but not after restarting the service (I have uploaded a screencast to the issue). Indeed, it seems the additional fix could help with this.

On the other hand, I still wonder if it makes sense that the empty capabilities overwrite the cached non-empty capabilities. But I guess, with the other fix this won't be a problem.

Reply:
@thebearon did you use browser refresh button? I think I saw in one of your videos you use 'Files' icon from the top menubar which isn't the same in result as browser refresh I think. (or am I wrong?)

Reply:
@juliushaertl can you implement the mentioned fix on the client side? I did short research and I wasn't able to get it working. It seems for me `OC.getCapabilities()` uses only some field generated in the html, I don't know how to fetch capabilities on demand.

Reply:
@eszkadev i usually click on the Files icon, or sometimes the folder icon to get out of the stalled loading screen.